go/ql/src/InconsistentCode/UnhandledCloseWritableHandle.ql

```diff
@@ -0,0 +1,160 @@
+  CloseFileFun() { this.hasQualifiedName("os.File", "Close") }
+}
+
+/**
+ * The `os.File.Sync` function.
+ */
+class SyncFileFun extends Function {
+  SyncFileFun() { this.hasQualifiedName("os.File", "Sync") }
+}
+
+/**
+ * Determines whether a call to a function is "unhandled". That is, it is either
+ * deferred or its result is not assigned to anything.
+ * TODO: maybe we should check that something is actually done with the result
+ */
+predicate unhandledCall(DataFlow::CallNode call) {
+  exists(DeferStmt defer | defer.getCall() = call.asExpr()) or
+  exists(ExprStmt stmt | stmt.getExpr() = call.asExpr())
+}
+
+/**
+ * Determines whether `source` is a writable file handle returned by a `call` to the
+ * `os.OpenFile` function.
+ */
+predicate isWritableFileHandle(DataFlow::Node source, DataFlow::CallNode call) {
+  exists(OpenFileFun f, DataFlow::Node flags, QualifiedName flag |
+    // check that the source is a result of the call
+    source = call.getAResult() and
+    // find a call to the os.OpenFile function
+    f.getACall() = call and
+    // get the flags expression used for opening the file
+    call.getArgument(1) = flags and
+    // extract individual flags from the argument
+    // flag = flag.getAChild*() and
+    flag = getConstants(flags.asExpr()) and
+    // check for one which signals that the handle will be writable
+    // note that we are underestimating here, since the flags may be
+    // specified elsewhere
+    isWritable(flag.getTarget())
+  )
+}
+
+/**
+ * Determines whether `os.File.Close` is called on `sink`.
+ */
+predicate isCloseSink(DataFlow::Node sink, DataFlow::CallNode call) {
+  exists(CloseFileFun f |
+    // find calls to the os.File.Close function
+    f.getACall() = call and
+    // that are deferred
+    unhandledCall(call) and
+    // where the function is called on the sink
+    call.getReceiver() = sink
+  )
+}
+
+/**
+ * Determines whether `os.File.Sync` is called on `sink` and the result of the call is neither
+ * deferred nor discarded.
+ */
+predicate isHandledSync(DataFlow::Node sink, DataFlow::CallNode syncCall) {
+  // find a call of the `os.File.Sync` function
+  syncCall = any(SyncFileFun f).getACall() and
+  // match the sink with the object on which the method is called
+  syncCall.getReceiver() = sink and
+  // check that the result is neither deferred nor discarded
+  not unhandledCall(syncCall)
+}
+
+/**
+ * A data flow configuration which traces writable file handles resulting from calls to
+ * `os.OpenFile` to `os.File.Close` calls on them.
+ */
+class UnhandledFileCloseDataFlowConfiguration extends DataFlow::Configuration {
+  UnhandledFileCloseDataFlowConfiguration() { this = "UnhandledCloseWritableHandle" }
+
+  override predicate isSource(DataFlow::Node source) { isWritableFileHandle(source, _) }
+
+  override predicate isSink(DataFlow::Node sink) { isCloseSink(sink, _) }
+}
+
+/**
+ * Determines whether a `DataFlow::CallNode` is preceded by a call to `os.File.Sync`.
```

Comment:
What if there are two control flow paths to the callnode, and one has a handled sync call and the other doesn't? Should this predicate hold or not? Currently it does. If you'd rather it didn't then you could maybe look into `dominates` in `BasicBlocks.qll`.